File: 백준/Gold/16236. 아기 상어/아기 상어.py
import sys 
print = lambda *args, sep=" ", end="\n": sys.stdout.write(sep.join(map(str, args)) + end)
input = lambda: sys.stdin.readline().rstrip('\r\n')

from collections import deque
from math import inf

def main():
    n = int(input())
    board = []
    size = 2
    time = 0
    eat = 0
    for i in range(n):
        row = [*map(int, input().split())]
        board.append(row)
        for j in range(n):
            if row[j] == 9: 
                x, y = i, j
                board[i][j] = 0
    def getfish(x, y, size):
        dist = [[-1]*n for _ in [0]*n]
        dq = deque([(x, y)])
        dist[x][y] = 0
        can = []
        mind = inf
        while dq:
            rx, ry = dq.popleft()
            if dist[rx][ry] >= mind: continue
            for dx, dy in ((-1, 0), (0, -1), (0, 1), (1, 0)):
                nx, ny = rx+dx, ry+dy
                if 0 <= nx < n and 0 <= ny < n and dist[nx][ny] == -1:
                    if board[nx][ny] <= size:
                        dist[nx][ny] = dist[rx][ry]+1
                        if 0 < board[nx][ny] < size:
                            can.append((nx, ny))
                            mind = dist[nx][ny]
                        else:dq.append((nx, ny))
        if can: return min(can), mind
        else: return None
    while sh := getfish(x, y, size):
        (x, y), d = sh
        board[x][y] = 0
        time += d
        eat += 1
        if eat == size:
            size += 1
            eat = 0
    print(time)
    return
if __name__ == '__main__':
    main()

File: 백준/Gold/16236. 아기 상어/아기 상어.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using pii = pair<int, int>;
using ai3 = array<int, 3>;

const int dx[] = {-1, 0, 0, 1}, dy[] = {0, -1, 1, 0};
const int INF = 1e9;
int n;
vector<vector<int>> board;

ai3 getfish(int x, int y, int size){
    vector<vector<int>> dist(n, vector<int>(n, -1));
    queue<pii> q; q.push({x, y});
    dist[x][y] = 0;
    vector<ai3> can;
    int mind = INF;
    while(q.size()){
        auto [rx, ry] = q.front(); q.pop();
        if(dist[rx][ry] >= mind) continue;
        for(int dir = 0; dir < 4; dir++){
            int nx = rx+dx[dir], ny = ry+dy[dir];
            if(0 <= nx && nx < n && 0 <= ny && ny < n && dist[nx][ny] == -1){
                if(board[nx][ny] <= size){
                    dist[nx][ny] = dist[rx][ry]+1;
                    if(0 < board[nx][ny] && board[nx][ny] < size){
                        mind = dist[nx][ny];
                        can.push_back({mind, nx, ny});
                    }
                    else q.push({nx, ny});
                }
            }
        }
    }
    if(can.size()) return *min_element(can.begin(), can.end());
    return {-1, -1, -1};
}

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    cin >> n;
    board.resize(n, vector<int>(n));
    int size = 2, time = 0, eat = 0;
    int x, y;
    for(int i = 0; i < n; i++) for(int j = 0; j < n; j++){
        cin >> board[i][j];
        if(board[i][j] == 9){ 
            x = i, y = j;
            board[i][j] = 0;
        }
    }
    while(1){
        ai3 sh = getfish(x, y, size);
        if(sh[0] == -1) break;
        auto [d, nx, ny] = sh;
        time += d;
        eat++;
        if(eat == size){
            size++;
            eat = 0;
        }
        board[nx][ny] = 0;
        x = nx, y = ny;
    }
    cout << time;
    return 0;
}